vite-plugin-cjs-interop/src/index.ts

```diff
@@ -61,53 +70,119 @@ export function cjsInterop(options: CjsInteropOptions): Plugin {
 			});
 
 			const toBeFixed: any[] = [];
-			const dynamicImportsToBeFixed: any[] = [];
 			const preambles: string[] = [];
+			let hasDynamicImportsToFix = false;
 
 			const { walk } = await walker;
 
 			walk(ast as any, {
 				enter(node) {
-					if (node.type === "ImportDeclaration") {
-						if (matchesDependencies(node.source.value as string)) {
-							toBeFixed.push(node);
-						}
-					} else if (node.type === "ImportExpression") {
-						if (
-							node.source.type === "Literal" &&
-							matchesDependencies(node.source.value as string)
-						) {
-							dynamicImportsToBeFixed.push(node);
-						}
+					switch (node.type) {
+						case "ImportDeclaration":
+						case "ExportNamedDeclaration":
+						case "ImportExpression":
+							if (
+								node.source &&
+								node.source.type === "Literal" &&
+								matchesDependencies(node.source.value as string)
+							) {
+								toBeFixed.push(node);
+								if (node.type === "ImportExpression") {
+									hasDynamicImportsToFix = true;
+								}
+							}
+							break;
+						default:
 					}
 				},
 			});
 
-			if (
-				toBeFixed.length === 0 &&
-				dynamicImportsToBeFixed.length === 0
-			) {
+			if (toBeFixed.length === 0) {
 				return;
 			}
 			const bottomUpToBeFixed = toBeFixed.reverse();
 
 			const ms = sourcemaps ? new MagicString(code) : null;
 			let counter = 1;
+			let specifierCounter = 1;
 			let isNamespaceImport = false;
 
-			for (const node of dynamicImportsToBeFixed.reverse()) {
-				const insertion = ".then(__cjs_dyn_import__)";
-				if (sourcemaps) {
-					ms!.appendRight(node.end, insertion);
-				} else {
-					code =
-						code.slice(0, node.end) +
-						insertion +
-						code.slice(node.end);
+			for (const node of bottomUpToBeFixed) {
+				if (node.type === "ImportExpression") {
+					const insertion = ".then(__cjs_dyn_import__)";
+					if (sourcemaps) {
+						ms!.appendRight(node.end, insertion);
+					} else {
+						code =
+							code.slice(0, node.end) +
+							insertion +
+							code.slice(node.end);
+					}
+					continue;
+				}
```

Comment:
This code was just moved. This is a slight behavior change; We no longer process dynamicImportsToBeFixed before bottomUpToBeFixed. Instead, we process all fixes in the reverse order that they are found in the module... Which seemed actually safer to me. But please review carefully.